File: src/系统/02．物品系统/08．装备移速.lua
local ____lualib = require("lualib_bundle")
local __TS__StringTrim = ____lualib.__TS__StringTrim
local ____exports = {}
local fourCCToStringCompat, normalizeMovespeed2, getMaxMovespeed2Info, jass, GetItemTypeId, GetUnitDefaultMoveSpeed, itemsData, R2I, stringChar
function fourCCToStringCompat(four)
    local c1 = stringChar(four % 256)
    local c2 = stringChar(R2I(four / 256) % 256)
    local c3 = stringChar(R2I(four / 65536) % 256)
    local c4 = stringChar(R2I(four / 16777216) % 256)
    return ((tostring(c4) .. tostring(c3)) .. tostring(c2)) .. tostring(c1)
end
function normalizeMovespeed2(unit, value)
    if not (value > 0) then
        return 0
    end
    if value < 1 then
        return GetUnitDefaultMoveSpeed(unit) * value
    end
    return value
end
function getMaxMovespeed2Info(unit, ignoreItem)
    local max = 0
    local name = ""
    local count = 0
    do
        local slot = 0
        while slot <= 5 do
            do
                local item = jass.UnitItemInSlot(unit, slot)
                if not item then
                    goto __continue10
                end
                if ignoreItem and item == ignoreItem then
                    goto __continue10
                end
                local tid = GetItemTypeId(item)
                local idStr = fourCCToStringCompat(tid)
                local entry = itemsData[idStr]
                local typ = entry and entry.type
                if typ == "任务" or typ == "药剂" or typ == "食品" then
                    goto __continue10
                end
                local v = entry and entry.movespeed2
                if type(v) == "number" and v > 0 then
                    count = count + 1
                end
                if type(v) == "number" and v > max then
                    max = v
                    name = (entry and entry.name) ~= nil and __TS__StringTrim(tostring(entry.name)) or "" or "未知"
                end
            end
            ::__continue10::
            slot = slot + 1
        end
    end
    return {value = max, name = name, count = count}
end
jass = require("jass.common")
GetItemTypeId = jass.GetItemTypeId
GetUnitDefaultMoveSpeed = jass.GetUnitDefaultMoveSpeed
local ____require_result_0 = require("系统.00．核心系统.01．事件中心.04．物品事件中心")
local onItemPickup = ____require_result_0.onItemPickup
local onItemDrop = ____require_result_0.onItemDrop
itemsData = require("系统.02．物品系统.01．装备数据").default
local ____require_result_1 = require("lib.扩展函数.Star扩展函数.00．SGSS")
local SGSS_SetState = ____require_result_1.SGSS_SetState
local ____require_result_2 = require("lib.扩展函数.BJ函数.08．单位BJ扩展")
local IsUnitIllusionBJ = ____require_result_2.IsUnitIllusionBJ
R2I = jass.R2I
stringChar = string.char
--- 单位已应用的 movespeed2 值（仅用于 SGSS 先减后加）
local applied = {}
local EQUIP_SPEED_EVENT_PLAYER_IDS = {
    0,
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    13
}
local function getUnitKey(unit)
    return tostring(unit)
end
local function getMaxMovespeed2(unit, ignoreItem)
    local info = getMaxMovespeed2Info(unit, ignoreItem)
    return normalizeMovespeed2(unit, info.value)
end
local function applyMovespeed2(unit, newSpeed)
    local key = getUnitKey(unit)
    local oldSpeed = applied[key] ~= nil and applied[key] or 0
    if newSpeed == oldSpeed then
        return
    end
    if oldSpeed ~= 0 then
        SGSS_SetState(unit, 9, -oldSpeed)
    end
    if newSpeed ~= 0 then
        SGSS_SetState(unit, 9, newSpeed)
    end
    applied[key] = newSpeed
end
local function onItemChange(unit, item, isPickup)
    if unit == nil or unit == 0 then
        return
    end
    if jass.IsUnitType(unit, jass.UNIT_TYPE_SUMMONED) then
        return
    end
    if IsUnitIllusionBJ(nil, unit) then
        return
    end
    local isDrop = not isPickup
    local newSpeed = isDrop and getMaxMovespeed2(unit, item) or getMaxMovespeed2(unit)
    local key = getUnitKey(unit)
    local cur = applied[key] ~= nil and applied[key] or 0
    if isPickup and newSpeed <= cur then
        return
    end
    applyMovespeed2(unit, newSpeed)
end
local function init()
    onItemPickup(function(unit, item)
        onItemChange(unit, item, true)
    end)
    onItemDrop(function(unit, item)
        onItemChange(unit, item, false)
    end)
end
init()
____exports.getMaxMovespeed2Info = getMaxMovespeed2Info
return ____exports

File: TS/系统/02．物品系统/08．装备移速.ts
/** @noSelfInFile */

/**
 * 装备移速（movespeed2）：不叠加，取当前装备中 movespeed2 最大值
 * 直接调用 TS 端 SGSS_SetState(unit, 9, value)；减速需传负数，故先减掉旧值再加新值。
 */
const jass = require("jass.common") as JassCommon;
const GetItemTypeId = (jass as any).GetItemTypeId as (this: void, item: any) => number;
const GetUnitDefaultMoveSpeed = (jass as any).GetUnitDefaultMoveSpeed as (this: void, unit: any) => number;
const { onItemPickup, onItemDrop } = require("系统.00．核心系统.01．事件中心.04．物品事件中心") as {
  onItemPickup: (this: void, callback: (this: void, unit: any, item: any) => void) => number;
  onItemDrop: (this: void, callback: (this: void, unit: any, item: any) => void) => number;
};
const itemsData = (require("系统.02．物品系统.01．装备数据") as { default: Record<string, { movespeed2?: number; name?: string; type?: string }> }).default;
const { SGSS_SetState } = require("lib.扩展函数.Star扩展函数.00．SGSS") as {
  SGSS_SetState: (this: void, u: any, id: number, v: number) => void;
};
const { IsUnitIllusionBJ } = require("lib.扩展函数.BJ函数.08．单位BJ扩展") as {
  IsUnitIllusionBJ: (unit: any) => boolean;
};
const R2I: any = (jass as any).R2I;
const stringChar: any = string.char;

/** 单位已应用的 movespeed2 值（仅用于 SGSS 先减后加） */
const applied: Record<string, number> = {};
const EQUIP_SPEED_EVENT_PLAYER_IDS = [0, 1, 2, 3, 4, 5, 6, 7, 13] as const;

function getUnitKey(unit: any): string {
  // Lua 环境 tostring(handle) 是稳定且唯一的（如 "unit:0x..."）
  return tostring(unit);
}

function fourCCToStringCompat(four: number): string {
  const c1 = stringChar(four % 256);
  const c2 = stringChar(R2I(four / 256) % 256);
  const c3 = stringChar(R2I(four / 65536) % 256);
  const c4 = stringChar(R2I(four / 16777216) % 256);
  return `${String(c4)}${String(c3)}${String(c2)}${String(c1)}`;
}

function getMaxMovespeed2(unit: any, ignoreItem?: any): number {
  const info = getMaxMovespeed2Info(unit, ignoreItem);
  return normalizeMovespeed2(unit, info.value);
}

function normalizeMovespeed2(unit: any, value: number): number {
  if (!(value > 0)) return 0;
  if (value < 1) return GetUnitDefaultMoveSpeed(unit) * value;
  return value;
}

/** 返回当前生效的移速值、提供该移速的装备名、以及带移速的装备件数（≥2 时才提示“只生效某装备”） */
function getMaxMovespeed2Info(unit: any, ignoreItem?: any): { value: number; name: string; count: number } {
  let max = 0;
  let name = "";
  let count = 0;
  for (let slot = 0; slot <= 5; slot++) {
    const item = (jass as any).UnitItemInSlot(unit, slot);
    if (!item) continue;
    if (ignoreItem && item === ignoreItem) continue;
    const tid = GetItemTypeId(item) as number;
    const idStr = fourCCToStringCompat(tid);
    const entry = (itemsData as Record<string, { movespeed2?: number; type?: string; name?: string }>)[idStr];
    const typ = entry?.type;
    if (typ === "任务" || typ === "药剂" || typ === "食品") continue;
    const v = entry?.movespeed2;
    if (typeof v === "number" && v > 0) count++;
    if (typeof v === "number" && v > max) {
      max = v;
      name = (entry?.name != null ? String(entry.name).trim() : "") || "未知";
    }
  }
  return { value: max, name, count };
}

function applyMovespeed2(unit: any, newSpeed: number): void {
  const key = getUnitKey(unit);
  const oldSpeed = applied[key] != null ? applied[key] : 0;
  if (newSpeed === oldSpeed) return;
  if (oldSpeed !== 0) {
    SGSS_SetState(unit, 9, -oldSpeed);
  }
  if (newSpeed !== 0) {
    SGSS_SetState(unit, 9, newSpeed);
  }
  applied[key] = newSpeed;
}

function onItemChange(unit: any, item: any, isPickup: boolean): void {
  if (unit === null || unit === 0) return;
  if (jass.IsUnitType(unit, (jass as any).UNIT_TYPE_SUMMONED)) return;
  if (IsUnitIllusionBJ(unit)) return;
  const isDrop = !isPickup;
  const newSpeed = isDrop ? getMaxMovespeed2(unit, item) : getMaxMovespeed2(unit);
  const key = getUnitKey(unit);
  const cur = applied[key] != null ? applied[key] : 0;
  if (isPickup && newSpeed <= cur) return;
  applyMovespeed2(unit, newSpeed);
}

function init(): void {
  // 使用物品事件中心注册，减少触发器数量
  onItemPickup((unit, item) => {
    onItemChange(unit, item, true);
  });
  onItemDrop((unit, item) => {
    onItemChange(unit, item, false);
  });
}

init();
/** 供装备系统在「当前装备加成」里显示移速：返回当前生效的移速值及提供该移速的装备名 */
export { getMaxMovespeed2Info };
